### PR TITLE
[xDS] Send notification to watchers in case last update was invalid and we receive identical update 

### PR DESCRIPTION
**Pull request by pawbhard**
Handles edge case described in #38094

Send notifications to watcher for identical cache object if last update was invalid.

### CODE REVIEW

File: test/core/xds/xds_client_test.cc [hunk 1]
```diff
@@ -6021,6 +6021,127 @@ TEST_F(XdsClientTest, FallbackOnStartup) {
...
 
+// flow steps :
+// 1. we  get valid resource
+// 2. we recieved invalid resource
+// 3. we receive valid resource again
+// Test that we redeliver resources to watcher
+// so they know error is cleared
+TEST_F(XdsClientTest, RedeliverSameResourceAfterError) {
+  InitXdsClient();
+
```
Comment (markdroth) [inline test/core/xds/xds_client_test.cc:6032]: Nit, for future reference: Please remove blank lines within functions.  In general, my feeling is that if you really need to break up sections of the code into logical chunks, you're better off adding a one-line comment indicating what the next chunk is doing rather than just having a blank line, since that helps readers understand the intent of the code.  Conversely, if something is really so obvious that it doesn't need the one-line comment, then it probably doesn't need to be broken up into a separate section anyway.

File: test/core/xds/xds_client_test.cc [hunk 2]
```diff
@@ -6021,6 +6021,127 @@ TEST_F(XdsClientTest, FallbackOnStartup) {
...
+          .set_nonce("B")
+          .AddInvalidResource(XdsFooResourceType::Get()->type_url(),
+                              "{\"name\":\"foo1\",\"value\":[]}")
+          .Serialize());
+
+  // XdsClient will ignore the error, so watcher should not see any event.
+  EXPECT_FALSE(watcher->HasEvent());
+
+  // XdsClient should deliver an error to the watcher.
+  auto error = watcher->WaitForNextAmbientError();
```
Comment (markdroth) [inline test/core/xds/xds_client_test.cc:6082]: If the error is delivered to the watcher via `OnAmbientError()` instead of `OnResourceChanged()`, then this isn't really the case we care about.

The watcher is expected to keep using the last thing that was passed to `OnResourceChanged()`, even if `OnAmbientError()` is called afterward.  The problem case that we were trying to solve here is the following sequence of events:

1. We deliver the initial valid resource to `OnResourceChanged()`.
2. We NACK and update and deliver the error to `OnResourceChanged()`.  At this point, the watcher stops using the resource delivered in the previous step.
3. We get a valid resource again.  We should deliver this to `OnResourceChanged()`, so that the watcher will once again start using the resource.

The problem that I described in the bug is that our current logic will inhibit the update in step 3 if the resource is identical to the cached version, which would incorrectly leave the watcher using the error instead of the resource.

In order to make the NACK error be reported to `OnResourceChanged()` instead of `OnAmbientError()`, I think we need to enable the `fail_on_data_errors` server feature.  However... writing that made me realize that there wouldn't actually be a bug in that case, because if we enable `fail_on_data_errors`, then we'll drop the cached resource, which means that the duplicate check won't actually inhibit the notification in step 3 above.

In other words, I think there isn't actually a bug here, and this PR isn't needed at all.  I'm sorry to have made you do work for nothing!  But hopefully it was still a good introduction to the code.

Incidentally, note that if we do deliver the error to `OnAmbientError()` instead of `OnResourceChanged()` in step 2 above, then there's no need to call `OnResourceChanged()` again with the valid resource, because the watcher will never have stopped using the resource it was given in step 1.  The only thing we need to do in that case is to call `OnAmbientError()` again with OK status to indicate that the ambient error has been cleared, and that case is already tested in the [`ConnectionFailsWithCachedResource` test case](https://github.com/grpc/grpc/blob/780a9df654e8d038cb07c353e8d212ab5331ad88/test/core/xds/xds_client_test.cc#L3713).

Sorry for sending you on a wild goose chase!
